src/states/node.rs

```diff
@@ -971,63 +1038,131 @@ impl Node {
-                nonce_bytes: nonce.0,
-                public_id: *self.full_id().public_id(),
-            }
-            .into_msg();
-
-        let request_msg = RoutingMessage {
-            src: src,
-            dst: dst,
-            content: request_content,
-        };
-
-        if let Err(err) = self.send_routing_message(request_msg) {
-            debug!("{:?} Failed to send connection info for {:?}: {:?}.",
-                   self,
-                   pub_id.name(),
-                   err);
         }
     }
 
-    // TODO: check whether these two methods can be merged into one.
-    fn handle_connection_info_from_client(&mut self,
-                                          conn_info: ConnectionInfo,
-                                          src: Authority<XorName>,
-                                          dst: Authority<XorName>)
-                                          -> Result<(), RoutingError> {
-        self.peer_mgr.allow_connect(conn_info.public_id.name())?;
-        self.connect(conn_info, dst, src)
+    fn handle_connection_info_request(&mut self,
+                                      encrypted_connection_info: Vec<u8>,
+                                      nonce_bytes: [u8; box_::NONCEBYTES],
+                                      public_id: PublicId,
+                                      message_id: MessageId,
+                                      src: Authority<XorName>,
+                                      dst: Authority<XorName>)
+                                      -> Result<(), RoutingError> {
+        let name = match src {
+            Authority::Client { .. } => public_id.name(),
+            Authority::ManagedNode(ref name) => name,
+            _ => unreachable!(),
+        };
+        self.peer_mgr.allow_connect(name)?;
+        let their_connection_info = self.decrypt_connection_info(&encrypted_connection_info,
+                                     &box_::Nonce(nonce_bytes),
+                                     &public_id)?;
+        let peer_id = their_connection_info.id();
+        use peer_manager::ConnectionInfoReceivedResult::*;
+        match self.peer_mgr
+            .connection_info_received(src, dst, public_id, their_connection_info, message_id) {
+            Ok(Ready(our_info, their_info)) => {
+                info!("{:?} Already sent a connection info request to {:?} ({:?}); resending our \
+                      same details as a response.",
+                      self,
+                      public_id.name(),
+                      peer_id);
+                let _ = self.do_send_connection_info(our_info.to_pub_connection_info(),
+                                                     public_id,
+                                                     dst,
+                                                     src,
+                                                     Some(message_id));
+                if let Err(error) = self.crust_service.connect(our_info, their_info) {
```

Comment:
is it necessary to carry out this connect attempt on receiving a `connection_info_request`?
I though a connect attempt shall only be made in case of receiving a `connection_info_response`?